File: C/Linked List/Singly Linked List/SinglyLL006.c
/*
=====================================================================

Program Name    : Singly Linked List - Multi Node Traversal
Program Number  : 006
Description     : This program demonstrates the creation and linking
                  of three nodes in a singly linked list. It shows
                  how node addresses are stored in pointers and how
                  multiple nodes can be accessed through chained
                  pointer traversal using the arrow (->) operator.

Concepts Used   : Structure, Self-Referential Structure,
                  Node Linking, Pointer Traversal,
                  Arrow Operator, Memory Addresses

Input           : None

Output          : Displays node addresses, pointer values,
                  and data stored in all linked nodes.

Date            : July 2026
Language        : C
Author          : Atharv Tushar Bhosale

=====================================================================
*/

#include<stdio.h>

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

int main()
{
    struct node obj1, obj2, obj3;

    obj1.data = 11;
    obj1.next = &obj2;

    obj2.data = 21;
    obj2.next = &obj3;

    obj3.data = 51;
    obj3.next = NULL;

    printf("%u\n", &obj1);
    printf("%u\n", &obj2);
    printf("%u\n", &obj3);

    printf("%u\n", obj1.next);
    printf("%u\n", obj2.next);
    printf("%u\n", obj3.next);

    printf("%d\n", obj1.data);
    printf("%d\n", obj1.next->data);
    printf("%d\n", obj1.next->next->data);

    return 0;
}

/*
=====================================================================

Expected Output :
(Address of obj1)
(Address of obj2)
(Address of obj3)

(Address stored in obj1.next)
(Address stored in obj2.next)
0

11
21
51

Explanation :
Three nodes are created and linked together to form
a simple singly linked list. The first node points
to the second node, and the second node points to
the third node. The third node contains NULL in its
next pointer, indicating the end of the list.

The program demonstrates chained traversal:

obj1.data                  -> Accesses first node data
obj1.next->data            -> Accesses second node data
obj1.next->next->data      -> Accesses third node data

Operations Performed :
1. Created three node objects.
2. Initialized data members.
3. Linked nodes sequentially.
4. Displayed node addresses.
5. Displayed pointer values.
6. Traversed multiple nodes using pointers.
7. Accessed data from each linked node.

Conclusion :
This program demonstrates multi-node traversal in a
singly linked list and shows how nodes can be
accessed sequentially through chained pointer
references.

Note :
For portable code on modern systems, pointers should
be printed using %p instead of %u.

=====================================================================
*/

File: Java/SinglyLL006.java
/*
=====================================================================

Program Name    : Singly Linked List Traversal using While Loop (Java)
Program Number  : 006
Description     : This program demonstrates traversal of a
                  Singly Linked List using a temporary
                  reference variable and a while loop.
                  All node elements are displayed
                  sequentially from the first node
                  to the last node.

Concepts Used   : Java, Class, Object,
                  Reference Variables,
                  Self-Referential Class,
                  Linked List Traversal,
                  While Loop

Input           : None

Output          : Displays all elements present
                  in the linked list.

Date            : July 2026
Language        : Java
Author          : Atharv Tushar Bhosale

=====================================================================
*/

class node
{
    public int data;
    public node next;
}

class SinglyLL006
{
    public static void main(String A[])
    {
        node head = null;
        node obj1 = null;
        node obj2 = null;
        node obj3 = null;

        obj1 = new node();
        obj2 = new node();
        obj3 = new node();

        obj1.data = 11;
        obj2.data = 21;
        obj3.data = 51;

        obj1.next = obj2;
        obj2.next = obj3;
        obj3.next = null;

        head = obj1;

        node temp = head;

        while(temp != null)
        {
            System.out.println(temp.data);
            temp = temp.next;
        }

    }
}

/*
=====================================================================

Expected Output :

11
21
51

Explanation :

Three node objects are created dynamically.

obj1 contains data 11
obj2 contains data 21
obj3 contains data 51

Links are established as:

head
 |
 v

+----+------+     +----+------+     +----+------+
| 11 |   o----->  | 21 |   o----->  | 51 | NULL |
+----+------+     +----+------+     +----+------+

A temporary reference variable temp is
initialized with the address stored in head.

Traversal Process :

Iteration 1:
temp -> 11
Display 11
Move temp to next node

Iteration 2:
temp -> 21
Display 21
Move temp to next node

Iteration 3:
temp -> 51
Display 51
Move temp to next node

Iteration 4:
temp becomes null
Loop terminates

Operations Performed :

1. Created three node objects.
2. Stored data in each node.
3. Linked all nodes together.
4. Assigned head to first node.
5. Created temporary reference temp.
6. Traversed linked list using while loop.
7. Displayed all node elements.

Conclusion :

The while loop successfully traverses
the Singly Linked List from the first
node to the last node and displays
all stored values.

=====================================================================
*/
